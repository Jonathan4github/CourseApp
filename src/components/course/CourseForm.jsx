import React from 'react';
import PropTypes from 'prop-types';

import TextInput from '../common/TextInput';
import SelectInput from '../common/SelectInput';


const propTypes = {
  course: PropTypes.shape().isRequired,
  allAuthors: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleOnSave: PropTypes.func.isRequired,
  handleOnChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  errors: PropTypes.shape()
};

const defaultProps = {
  errors: {}
};

const CourseForm = ({
  course, allAuthors, handleOnSave, handleOnChange, isLoading, errors
}) => (
  <form>
    <TextInput
      name="title"
      label="Title"
      value={course.title}
      onChange={handleOnChange}
      error={errors.title} />
    <SelectInput
      name="authorId"
      label="Author"
      value={course.authorId}
      defaultOption="Select Author"
      options={allAuthors}
      onChange={handleOnChange}
      error={errors.authorId} />
    <TextInput
      name="category"
      label="category"
      value={course.category}
      onChange={handleOnChange}
      error={errors.category} />
    <TextInput
        name="length"
        label="length"
        value={course.length}
        onChange={handleOnChange}
        error={errors.length} />
    <input
     type="submit"
     className="btn btn-primary"
     value={isLoading ? 'Saving' : 'Save'}
     onClick={handleOnSave} />
  </form>
);

CourseForm.propTypes = propTypes;
CourseForm.defaultProps = defaultProps;

export default CourseForm;
