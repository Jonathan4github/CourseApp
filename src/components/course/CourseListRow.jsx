import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';


const propTypes = {
  course: PropTypes.shape().isRequired,
  handleDelete: PropTypes.func.isRequired
};

const CourseListRow = ({ course, handleDelete }) => (
  <tr>
    <td>
      <a href={course.watchHref} target="_blank" rel="noopener noreferrer">Watch</a>
    </td>
    <td><Link to={`/course/${course.id}`}>{course.title}</Link></td>
    <td>{course.authorId}</td>
    <td>{course.category}</td>
    <td>{course.length}</td>
    <td>
      <button type="button" onClick={() => handleDelete(course.id)} className="btn btn-danger">Delete</button>
    </td>
  </tr>
);

CourseListRow.propTypes = propTypes;

export default CourseListRow;
