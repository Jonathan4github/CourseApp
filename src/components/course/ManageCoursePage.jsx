import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Prompt } from 'react-router-dom';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import Loader from 'react-md-spinner';

import CourseForm from './CourseForm';
import NotFoundPage from '../common/NotFoundPage';
import * as courseActions from '../../actions/creators/courseActions';
import * as authorActions from '../../actions/creators/authorActions';

import { validateFormData } from '../../helpers/validations';


const propTypes = {
  actions: PropTypes.shape({
    loadSingleCourse: PropTypes.func.isRequired,
    saveCourse: PropTypes.func.isRequired
  }).isRequired,
  course: PropTypes.shape({
    selectedCourse: PropTypes.shape(),
    isFetching: PropTypes.bool.isRequired,
    isSaving: PropTypes.bool.isRequired
  }),
  allAuthors: PropTypes.array.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape().isRequired
  }).isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired
};

const defaultProps = {
  course: PropTypes.shape({
    selectedCourse: {},
    isFetching: false,
    isSaving: false
  })
};

export class ManageCoursePage extends Component {
  match = this.props.match;

  history = this.props.history;

  initialFormState = {
    course: {
      id: '',
      watchHref: '',
      title: '',
      authorId: '',
      length: '',
      category: ''
    },
    errors: {},
    isBlocking: false
  };

  state = { ...this.initialFormState };

  componentDidMount() {
    const { actions } = this.props;

    if (this.match.params.id !== 'add') {
      actions.loadSingleCourse(this.match.params.id);
    }
  }

  componentWillReceiveProps(nextProps) {
    const { match, course: { selectedCourse } } = this.props;
    const { course } = nextProps;
    if (match.params.id !== 'add') {
      if (course.selectedCourse !== selectedCourse) {
        this.setState(prevState => ({
          ...prevState,
          course: nextProps.course.selectedCourse
        }));
      }
    } else {
      this.setState(prevState => ({
        ...prevState
      }));
    }
  }

  handleOnChange = (event) => {
    const { name, value } = event.target;
    this.setState(prevState => ({
      ...prevState,
      course: { ...prevState.course, [name]: value },
      isBlocking: true
    }));
  };

  isValid = (course) => {
    const { errors, isValid } = validateFormData(course);
    if (!isValid) {
      this.setState(() => ({
        errors
      }));
    }
    return isValid;
  }

  handleOnFocus = (event) => {
    event.persist();
    this.setState(prevState => ({
      errors: { ...prevState.errors, [event.target.name]: '' }
    }));
  }

  handleOnSave = (event) => {
    const { actions } = this.props;
    const { course } = this.state;

    event.preventDefault();

    if (this.isValid(course)) {
      actions.saveCourse(course)
        .then(() => {
          this.setState(prevState => ({ isBlocking: !prevState }), () => {
            this.history.push('/courses');
          });
        });
    }
  };

  render() {
    const { course, errors, isBlocking } = this.state;
    const { allAuthors, course: { isFetching, isSaving, selectedCourse } } = this.props;

    if (!isFetching && this.match.params.id && typeof selectedCourse === 'undefined') {
      return <NotFoundPage history={this.history} />;
    }

    return (
      <Fragment>
        <Prompt
          when={isBlocking}
          message={
            location => `Are you sure you want to go to ${location.pathname}`
          }
        />
        <div className="top-container">
          <h2>Manage Course</h2>
        </div>
        {isFetching
          ? (
            <div className="loader-container">
              <Loader size="30" className="loader-component" />
            </div>
          )
          : (
            <CourseForm
              allAuthors={allAuthors}
              course={course}
              handleOnChange={this.handleOnChange}
              handleOnSave={this.handleOnSave}
              handleOnFocus={this.handleOnFocus}
              isFetching={isFetching}
              isSaving={isSaving}
              errors={errors} />
          )
        }
      </Fragment>
    );
  }
}

const mapStateToProps = state => ({
  course: state.courses,
  allAuthors: state.authors.allAuthors
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ ...courseActions, ...authorActions }, dispatch)
});

ManageCoursePage.propTypes = propTypes;
ManageCoursePage.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(ManageCoursePage);
