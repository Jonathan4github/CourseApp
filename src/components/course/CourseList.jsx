import React from 'react';
import PropTypes from 'prop-types';
import CourseListRow from './CourseListRow';


const propTypes = {
  courses: PropTypes.array.isRequired,
  handleDelete: PropTypes.func.isRequired
};

const CourseList = ({ courses, handleDelete }) => (
  <table className="table table-hover">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>Title</th>
        <th>Author</th>
        <th>Category</th>
        <th>Length</th>
        <th>Actions</th>
      </tr>
    </thead>
    <tbody>
      {courses
        .map(course => (
          <CourseListRow
            key={course.id}
            course={course}
            handleDelete={handleDelete}
            />
        ))
      }
    </tbody>
  </table>
);

CourseList.propTypes = propTypes;

export default CourseList;
