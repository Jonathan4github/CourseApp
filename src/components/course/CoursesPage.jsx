import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Loader from 'react-md-spinner';

import CourseList from './CourseList';
import * as courseActions from '../../actions/creators/courseActions';

const propTypes = {
  courses: PropTypes.shape({
    allCourses: PropTypes.array.isRequired,
    isLoading: PropTypes.bool.isRequired
  }).isRequired,
  actions: PropTypes.shape().isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired
};

class CoursesPage extends Component {
  componentDidMount() {
    const { actions, courses: { allCourses } } = this.props;

    actions.loadCourses();
  }

  showAddCoursePage = () => {
    const { history } = this.props;
    history.push('/course');
  };

  courseRow = (course, index) => (
    <div key={index}>{course.title}</div>
  );

  render() {
    const { courses: { allCourses, isLoading } } = this.props;
    return (
      <div>
        <div className="top-container">
          <h1>Courses</h1>
          <button
            type="button"
            className="btn btn-outline-info add-course-button btn-sm"
            onClick={this.showAddCoursePage}>
            Add Course
          </button>
        </div>
        {isLoading
          ? (
            <div className="loader-container">
              <Loader size="30" className="loader-component" />
            </div>
          )
          : <CourseList courses={allCourses} />
        }
      </div>
    );
  }
}

const mapStateToProps = state => ({
  courses: state.courses
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(courseActions, dispatch)
});

CoursesPage.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(CoursesPage);
