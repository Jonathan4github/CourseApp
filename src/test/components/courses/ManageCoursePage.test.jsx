/* eslint-disable */
import { ManageCoursePage } from '../../../components/course/ManageCoursePage';
import coursesMock from '../../__mocks__/courseMock';


describe('<ManageCoursePage /> component', () => {
  const setup = () => {
    const props = {
      actions: {
        loadSingleCourse: jest.fn(),
        saveCourse: jest.fn()
      },
      course: {
        selectedCourse: {},
        isFetching: false,
        isSaving: false
      },
      allCourses: coursesMock.courses,
      match: {
        params: {
          id: 'add'
        }
      },
      history: {
        push: jest.fn()
      }
    };

    const state = {
      errors: {},
      isBlocking: false
    };

    return {
      pressentationWrapper: shallow(<ManageCoursePage {...props} />),
      props,
      state
    };
  };

  const event = {
    target: {
      name: 'title',
      value: 'some title'
    },
    preventDefault: jest.fn(),
    persist: jest.fn()
  };

  it('render component without crashing', () => {
    const { pressentationWrapper } = setup();

    expect(toJson(pressentationWrapper)).toMatchSnapshot();
  });

  it('invokes `componentDidMount` when mounted', () => {
    jest.spyOn(ManageCoursePage.prototype, 'componentDidMount');
    setup();

    expect(ManageCoursePage.prototype.componentDidMount).toHaveBeenCalled();
    ManageCoursePage.prototype.componentDidMount.mockRestore();
  });

  it('renders a <Prompt /> child component', () => {
    const { pressentationWrapper } = setup();
    expect(pressentationWrapper.find('Prompt').length).toBe(1);
  });

  it('should show <Prompt /> component if value of its child prop(when) is true', () => {
    const { pressentationWrapper, state } = setup();

    pressentationWrapper.setState({
      ...state, isBlocking: true
    });

    expect(pressentationWrapper.find('Prompt').props().when).toEqual(true);
  });

  it('invokes handleOnChange method', () => {
    const { pressentationWrapper } = setup();
    const handleOnChangeSpy = jest
      .spyOn(pressentationWrapper.instance(), 'handleOnChange');
    pressentationWrapper.instance().handleOnChange(event);

    expect(handleOnChangeSpy).toHaveBeenCalled();
  });

  it('invokes handleOnSave method', () => {
    const { pressentationWrapper } = setup();
    const handleOnSaveSpy = jest.spyOn(pressentationWrapper.instance(), 'handleOnSave');
    pressentationWrapper.instance().handleOnSave(event);
    expect(handleOnSaveSpy).toHaveBeenCalled();
  });

  it('invokes handleOnFocus method', () => {
    const { pressentationWrapper } = setup();
    const handleOnFocusSpy = jest.spyOn(pressentationWrapper.instance(), 'handleOnFocus');
    pressentationWrapper.instance().handleOnFocus(event);
    expect(handleOnFocusSpy).toHaveBeenCalled();
  });

  it('shows <Loader /> component when `isFetching` is true', () => {
    const { pressentationWrapper, props } = setup();
    pressentationWrapper.setProps({
      ...props,
      course: { ...props.course, isFetching: true }
    });

    expect(pressentationWrapper.find('.loader-container')).toHaveLength(1);
  });

  it('shows <NotFoundPage /> component when `courseId` does not exist', () => {
    const { pressentationWrapper, props } = setup();
    pressentationWrapper.setProps({
      ...props,
      course: { ...props.course, selectedCourse: undefined }
    });

    expect(pressentationWrapper.find('NotFoundPage')).toHaveLength(1);
  });
});
