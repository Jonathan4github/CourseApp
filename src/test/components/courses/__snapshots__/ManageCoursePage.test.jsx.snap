// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageCoursePage /> component render component without crashing 1`] = `
<Fragment>
  <Prompt
    message={[Function]}
    when={false}
  />
  <div
    className="top-container"
  >
    <h2>
      Manage Course
    </h2>
  </div>
  <CourseForm
    course={
      Object {
        "authorId": "",
        "category": "",
        "id": "",
        "length": "",
        "title": "",
        "watchHref": "",
      }
    }
    errors={Object {}}
    handleOnChange={[Function]}
    handleOnFocus={[Function]}
    handleOnSave={[Function]}
    isFetching={false}
    isSaving={false}
  />
</Fragment>
`;
